### DISCUSSION TITLE
Theming issues

### DISCUSSION THREAD

**Discussion by lsaavedr** [makeswift/makeswift#831] [Ideas]
Hello, I need to make colors and fonts available throughout the application, starting with the font, makeswift has a way to make this with a `handler` in the file `src/app/api/makeswift/[...makeswift]/route.ts` for example:

```typescript
const handler = MakeswiftApiHandler(process.env.MAKESWIFT_SITE_API_KEY, {
  runtime,
  getFonts() {
    return [
      {
        family: "Dancing Script",
        label: "Brand",
        variants: [
          { weight: "400", style: "normal" },
        ],
      },
    ]
  },
})
```
this works perfectly but I think it has a problem: if I set this font with the `Brand` label in MakeSwift it works, but if after that in the code I change the `family` it stops working, I need change in all the texts that have the `Brand` font the new `Brand` font. I think the label should have a higher priority than the family.

The other issue is in the use of CSS variables, for example with the default `handler`:
```typescript
const handler = MakeswiftApiHandler(process.env.MAKESWIFT_SITE_API_KEY, {
  runtime,
  getFonts() {
    return [
      {
        family: 'var(--font-inter)',
        label: 'Inter',
        variants: [
          { weight: '300', style: 'normal' },
          { weight: '300', style: 'italic' },
          { weight: '400', style: 'normal' },
          { weight: '400', style: 'italic' },
          { weight: '700', style: 'normal' },
          { weight: '800', style: 'normal' },
        ],
      },
    ]
  },
})
```
It suggests that we can use CSS variables to define the font family, It doesn't work, the font family will be the string ` 'var(--font-inter)'` it doesn't eval the CSS variable.

Regardless of these issues, I understand how to integrate app-wide fonts with the font set in makeswift, I love it, and I love the idea to make the same with the colors, set colors in the `handler` with labels, and then has these color-labels in makewift to compose the app.

**Comment by fikrikarim** [comment]
Hi @lsaavedr, thank you for the feedback.

We have a new approach for making colors available throughout the application by leveraging our new [`<MakeswiftComponent />` API ](https://docs.makeswift.com/developer/reference/components/makeswift-component)and creating a custom Site Theme component. You can see it in action [in this video](https://youtu.be/Mz5YE3UNoAw?si=EeotIKKP0Zhwe4tm&t=609).

Here's the [source code](https://github.com/bigcommerce/catalyst/blob/integrations/makeswift/core/lib/makeswift/components/site-theme/client.tsx) of the Site Theme component from our [Catalyst framework](https://www.catalyst.dev/). 

Summary of how to implement it:
- Create a `<SiteTheme />` component that accepts colors and fonts as props, transforms them as CSS variables, and then renders them inside a `<style />` tag.
- Register the site theme component and add [Color control](https://docs.makeswift.com/developer/reference/controls/color).
- Render the `<SiteTheme />` component inside a `<head />` tag in a `layout.tsx`. This makes the `<SiteTheme />` component available on all pages.
- Use the CSS variables on other components.

**Reply by lsaavedr** [reply]
Hi @fikrikarim ,

Thank you for your response and for sharing the details!

I checked the link to the video, but it seems to redirect me to the documentation instead of an actual video. Could you confirm if the link is correct?

Also, if you have the source code of a working example, that would be more than enough for me to understand the implementation.

I really appreciate your help!

**Reply by lsaavedr** [reply]
Oops, sorry! In my email, the link to the video was broken, but it's correct on GitHub. Thanks!

**Reply by fikrikarim** [reply]
>I checked the link to the video, but it seems to redirect me to the documentation instead of an actual video. Could you confirm if the link is correct?

Oh yeah, I updated the link a minute after I posted the message. Could you check it again?

Yes, here's an example [source code](https://github.com/bigcommerce/catalyst/blob/integrations/makeswift/core/lib/makeswift/components/site-theme/client.tsx).

**Reply by lsaavedr** [reply]
Hi @fikrikarim,

I’ve worked on the design approach you suggested, but unfortunately, it still has the same issue as the original approach. Additionally, it also introduces the added complexity of implementing a "SiteTheme," which makes it even more challenging.

The first point I want to clarify is that without implementing "SiteTheme," it’s still possible to choose colors from the color palette and fonts from the font set. When a color in the palette or a font in the set is changed, it updates everywhere automatically, just as shown in the video but without the need to implement "SiteTheme".

The second point is my perspective on the design: I don’t want to define colors and fonts in Makeswift. Instead, I want to define them in the code, just as it’s done with fonts [here](https://docs.makeswift.com/developer/reference/makeswift-api-handler#adding-fonts):

```typescript
const handler = MakeswiftApiHandler(process.env.MAKESWIFT_SITE_API_KEY, {
  runtime,
  getFonts() {
    return [
      {
        family: 'var(--font-inter)',
        label: 'Inter',
        variants: [
          { weight: '300', style: 'normal' },
          { weight: '300', style: 'italic' },
          { weight: '400', style: 'normal' },
          { weight: '400', style: 'italic' },
          { weight: '700', style: 'normal' },
          { weight: '800', style: 'normal' },
        ],
      },
    ]
  },
})
```

The reason for defining fonts and colors in the code, rather than in Makeswift GUI, is to use that code as the single source of truth for both pages managed by Makeswift and pages that are not managed by Makeswift.

Think of it as adding fonts, but in this case, I want to "add colors"—such as "primary-color," "secondary-color," "accent-color," etc.—directly in the code to ensure consistency across all pages (managed or not managet by Makeswift).

regards!

**Reply by fikrikarim** [reply]
Hi @lsaavedr,

You’re right, currently, Makeswift doesn’t yet support defining colors in code the same way fonts can be.

That said, this is a great idea and definitely the direction we want to move toward. We’ve been planning to support a code-based approach for colors similar to how `getFonts` works. We don’t have a timeline for this yet, but it’s on our radar.

In the meantime, if you want to manage colors in code, one option is to expose color options using a `Select` control in your custom components, though this won't work with the RichText.

**Comment by lsaavedr** [comment]
Hi @fikrikarim, another important aspect of the design is being able to handle dark mode. The colors are different in this mode, and my original idea was to use CSS variables and the @theme directive from TailwindCSS to manage this without having to handle it directly in Makeswift. However, since you’re planning to consider it, it would be ideal for Makeswift to also support dark mode. And, please, add a dark mode to the Makeswift interface—it’s hard on my eyes! 😅 Thanks!!!

**Reply by fikrikarim** [reply]
Hahah yeah, dark mode support has been on our roadmap for a while, but we still have some higher priority issues to tackle first. I forwarded this to the team.

There’s also a “Request a feature” section under the “Help” button. There’s already a dark mode request there that you can upvote 😄

<img width="1920" alt="image" src="https://github.com/user-attachments/assets/ce26cb27-5cc1-4191-bc6f-3acbcc88595e" />
